File: SRTI_Source/v1_v2_Java_Server_Client_Source/SRTI_v2_00_00/src/mainServer/RTISimConnectThread.java
package mainServer;

/*
 * RTISimConnectThread.java
 * 
 * - equivalent to "RTIConnectThread.java", but for RTILib and simulation side to handle dedicated thread for listening to new messages.
 * */

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;

public class RTISimConnectThread extends Thread{
	
	static int threadIndexTotal = 0;
	int threadIndex = 0;
	String tag = "RTISimConnectThread";
	String hostName = "localHost";
	int portNumber = 42;
	Socket rtiSocket;
	RTILib rtiLib;
	
	public RTISimConnectThread(RTILib rtiLib, Socket rtiSocket) {
		
		threadIndex = threadIndexTotal;
		threadIndexTotal++;
		
		this.rtiSocket = rtiSocket;
		this.rtiLib = rtiLib;
		
		Runtime.getRuntime().addShutdownHook(new Thread() {
			public void run() {
				try {
					printLine("you didn't call 'disconnect()' properly, but we sense that the JVM has closed, so we will force the connection closed.");
					isConnected = false;
				} catch (Exception e) {
					printLine("tried to force close the socket upon program ending, but something when wrong..." + e.toString());
				}
			}
		});
	}
	
	boolean isConnected = true;
	public void run() {
		printLine("Connected to dedicated socket! Now running thread to receive new messages!");
		BufferedReader in;
		try {
			in = new BufferedReader(new InputStreamReader(rtiSocket.getInputStream()));
			continuousInput(in);
		} catch (IOException e) {
			printLine(System.currentTimeMillis() + " Exception (some error trying to open the bufferedreader from the RTI...) >> " + e.toString());
		}
		
	}
	
	private long crashTimeInMillis = 0;
	private int numOfRTICrash = 0;
	private void continuousInput(BufferedReader in) {
		try {
			printLine("Connected to dedicated socket! Now running thread to receive new messages!");
			String userInput = in.readLine();
			String userInputOut = "(" + userInput.length() + ") " + userInput;
			if (userInput.length() > 1000) {
				userInputOut = userInput.substring(0,500) 
						+ " ... " + userInput.substring(userInput.length() - 500,userInput.length());
			}
			printLine("had read first line, received input = " + userInputOut);
			while (isConnected == true && userInput.length() > 0) {
				rtiLib.receivedMessage(userInput);
				printLine("ready to read...");
				userInput = in.readLine();
				userInputOut = "(" + userInput.length() + ") " + userInput;
				if (userInput.length() > 1000) {
					userInputOut = "(" + userInput.length() + ") " + userInput.substring(0,500) 
							+ " ... " + userInput.substring(userInput.length() - 500,userInput.length());
				}
				printLine("had read, received input = " + userInputOut);
			}
		} catch (Exception e) {
			disconnectedErr = System.currentTimeMillis() + " Exception (connection closed) >> " + e.toString();
			printLine(disconnectedErr);
			e.printStackTrace();
		}
		
		// if thread reading from RTI broke multiple times within 10 seconds, then there is some problem, maybe best to close the thread...
		if (System.currentTimeMillis() - crashTimeInMillis < 10000) {
			numOfRTICrash++;
		} else {
			numOfRTICrash = 0;
			numOfRTICrash++;
		}
		crashTimeInMillis = System.currentTimeMillis();
		
		if (isConnected == true) {
			if (numOfRTICrash >= 3) {
				printLine("Something serious is wrong with the connection to the RTI. Shut down the thread.");
				System.out.println("Something serious is wrong with the connection to the RTI. Shut down the thread.");
				rtiLib.reconnect();
			} else {
				continuousInput(in);
				return;
			}
		}
		
		try {
			in.close();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	private String disconnectedErr = "";
	public String getDisconnectedErr() {
		String returnString = null;
		if (disconnectedErr.length() > 0) {
			returnString = disconnectedErr;
			disconnectedErr = "";
		}
		return returnString;
	}
	
	public void closeConnection() {
		try {
			printLine("closing connection to RTI.");
			isConnected = false;
			rtiSocket.close();
		} catch (IOException e) {
			printLine("Error occurred when trying to close conenction to the RTI:" + e.toString());
		}
	}
	
	public void printLine(String line) {
		String formatLine = String.format("%1$32s", "[" + tag + "]" + " --- ") + line;
		Version.printSimConsole(formatLine);
		Version.printSimFile(formatLine);
		Version.printSimDebugGUI(formatLine);
	}
}

File: SRTI_Source/v2_Python_Sim_Wrapper_Source/Python/RTISimConnectThread.py

class RTISimConnectThread:

    threadIndexTotal = 0;
    threadIndex = 0;
    tag = "RTISimConnectThread";
    hostName = "localHost";
    portNumber = 42;
    rtiSocket = None;
    rtiLib = None;

    def __init__(self, rtiLib, rtiSocket):
        self.threadIndex = RTISimConnectThread.threadIndexTotal;
        RTISimConnectThread.threadIndexTotal = RTISimConnectThread.threadIndexTotal + 1;

        self.rtiLib = rtiLib;
        self.rtiSocket = rtiSocket;
    
    isConnected = True;
    def run(self):
        self.printLine("Connected to dedicated socket! Now running thread to receive new messages!");
        try:
            inReader = self.rtiSocket.makefile('r');
            self.continuousInput(inReader);
        except Exception as e:
            self.printLine(" Exception (some error trying to open the bufferedreader from the RTI...)");
            self.printLine(str(e));
            
    crashTimeInMillis = 0;
    numOfRTICrash = 0;
    def continuousInput(self, inp):
        try:
            self.printLine("Connected to dedicated socket! Now running thread to receive new messages!");
            from time import sleep;
            #sleep(1);

            userInput = inp.readline();
            #userInput = str(userInput);
            #self.printLine(userInput);
            #print("RECEIVED MESSAGE: ");
            #print(userInput);
            #print("THAT WAS RECEIVED MESSAGE.");
            #self.printLine(type(userInput).__name__);
            #print(type(userInput));
            #print(len(userInput));
            userInputOut = "(" + str(len(userInput)) + ") " + userInput;
            if (len(userInput) > 1000):
                userInputOut = "(" + str(len(userInput)) + ") " + userInput[:1000];
            self.printLine("had read first line, received input = " + userInputOut);
            while (self.isConnected == True and len(userInput) > 0):
                self.rtiLib.receivedMessage(userInput);
                self.printLine("ready to read...");
                userInput = inp.readline();
                userInput = str(userInput);
                userInputOut = "{" + str(len(userInput)) + ") " + userInput;
                if (len(userInput) > 1000):
                    userInputOut = "(" + str(len(userInput)) + ") " + userInput[:1000];
                self.printLine("had read, received input = " + userInputOut);
        except Exception as e:
            self.disconnectedErr = "... some error.";
            self.printLine(str(e));
            
        # normally assume crash if it occurs frequently within few seconds, but for simplicity, time limit is ignored in Python version.
        self.numofRTICrash = self.numOfRTICrash + 1;
        if (self.isConnected == True):
            if (self.numOfRTICrash >= 3):
                self.printLine("Something serious is wrong with the connection to the RTI. Shut down the thread.");
                self.rtiLib.reconnect();
            else:
                self.continuousInput(inp);
                return 0;

        try:
            inp.close();
        except Exception as e:
            self.printLine("Some last-minute error when trying to close the socket to the RTI Server.");
            self.printLine(str(e));
            
    disconnectedErr = "";
    def getDisconnectedErr(self):
        returnString = "";
        if (len(self.disconnectedErr) > 0):
            returnString = self.disconnectedErr;
            self.disconnectedErr = "";
        return returnString;
    
        
    def closeConnection(self):
        try:
            self.printLine("closing connection to RTI.");
            self.isConnected = False;
            self.rtiSocket.close();
        except:
            self.printLine("Error occurred when trying to close connection to the RTI.");
        return 0;



    def printLine(self, line):
        print("[RTISimConnectThread] " + line);
        return 0;
